homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,242 @@
+"""
+Support for Songpal-enabled (Sony) media devices.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.songpal/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    PLATFORM_SCHEMA, SUPPORT_SELECT_SOURCE, SUPPORT_TURN_OFF,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_STEP, SUPPORT_VOLUME_SET,
+    SUPPORT_TURN_ON, MediaPlayerDevice, DOMAIN)
+from homeassistant.const import (
+    CONF_NAME, STATE_ON, STATE_OFF, ATTR_ENTITY_ID)
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['python-songpal==0.0.6']
+
+SUPPORT_SONGPAL = SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP | \
+                  SUPPORT_VOLUME_MUTE | SUPPORT_SELECT_SOURCE | \
+                  SUPPORT_TURN_ON | SUPPORT_TURN_OFF
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SET_SOUND_SETTING = "songpal_set_sound_setting"
+PARAM_NAME = "name"
+PARAM_VALUE = "value"
+
+CONF_ENDPOINT = "endpoint"
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=None): cv.string,
+    vol.Required(CONF_ENDPOINT): cv.string,
+})
+
+
+async def async_setup_platform(hass, config,
+                               async_add_devices, discovery_info=None):
+    """Set up the Songpal platform."""
+    devices = []
+    if discovery_info is not None:
+        _LOGGER.debug("Got autodiscovered device: %s", discovery_info)
+        devices.append(
+            SongpalDevice(discovery_info["name"],
+                          discovery_info["properties"]["endpoint"]))
+    else:
+        songpal = SongpalDevice(config.get(CONF_NAME),
+                                config.get(CONF_ENDPOINT))
+        devices.append(songpal)
+
+    async_add_devices(devices, True)
+
+    async def async_service_handler(service):
+        """Service handler."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_id = service.data.get("entity_id", None)
+        _LOGGER.debug("Calling %s (entity: %s) with params %s",
+                      service, entity_id, params)
+
+        for device in devices:
```

Comment:
This won't work. `devices` is not shared between 2 configured entities. It would also mean that you are registering the service twice.

Reply:
Indeed, no idea where I came up with that idea. Converted it to store devices in `hass.data`.